required/latex-lab/latex-lab-mathcapture.dtx

```diff
@@ -0,0 +1,505 @@
+%
+% \MaybeStop{\setlength\IndexMin{200pt}  \PrintIndex  }
+%
+% \section{The Implementation}
+%
+%    \begin{macrocode}
+%<@@=math>
+%    \end{macrocode}
+%
+%    \begin{macrocode}
+%<*kernel>
+%    \end{macrocode}
+%
+% \subsection{File declaration}
+%    \begin{macrocode}
+\ProvidesFile{latex-lab-mathcapture.ltx}
+        [2023-01-06 v0.1a Grab all the math(s)]
+%    \end{macrocode}
+%
+%    \begin{macrocode}
+\ExplSyntaxOn
+%    \end{macrocode}
+%
+% \subsection{Setup}
+%
+% Loading \pkg{amsmath} is an absolute requirement: this avoids needing to
+% have conditional definitions and deals with how to define \cs{[}/\cs{]}
+% neatly.
+%    \begin{macrocode}
+\tl_gput_right:Nn \@kernel@before@begindocument
+  { \RequirePackage { amsmath } }
+%    \end{macrocode} 
+%
+% \subsection{Data structures}
+%
+% \begin{variable}{\l_@@_collected_bool}
+%   Tracks if math mode material has been collected, which happens inside
+%   \pkg{amsmath} environments as well as those handled directly here.
+%    \begin{macrocode}
+\bool_new:N \l_@@_collected_bool
+%    \end{macrocode}
+% \end{variable}
+%
+% \subsection{Interface commands}
+%
+% \begin{macro}
+%   {\@@_process:nn, \@@_process:Vn, \@@_process_auxi:nn, \@@_process_auxii:nn}
+%   A no-op place-holder; the internal wrapper means that it does not need to
+%   be concerned with internals.
+%    \begin{macrocode}
+\cs_new_protected:Npn \@@_process:nn #1#2
+  {
+    \legacy_if:nF { measuring@ }
+      { \tl_trim_spaces_apply:nN {#2} \@@_process_auxi:nn {#1} }
+  }
+\cs_generate_variant:Nn \@@_process:nn { V }
+\cs_new_protected:Npn \@@_process_auxi:nn #1#2
+  { \@@_process_auxii:nn {#2} {#1} }
+\cs_new_protected:Npn \@@_process_auxii:nn #1#2 { }
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}{\math_processor:n}
+%   A simple installer
+%    \begin{macrocode}
+\cs_new_protected:Npn \math_processor:n #1
+  { \cs_set_protected:Npn \@@_process_auxii:nn ##1##2 {#1} }
+%    \end{macrocode}
+% \end{macro}
+%
+% \subsection{Content grabbing}
+%
+% \begin{macro}{\@@_grab_dollar:w}
+%   Grab up to a single |$|, for inline math mode, suppressing
+%   any processing if the first token is \tn{m@th}.
+%    \begin{macrocode}
+\cs_new_protected:Npn \@@_grab_dollar:w % $
+  #1 $
+  {
+    \tl_if_head_eq_meaning:nNF {#1} \m@th
+      { \@@_process:nn { math } {#1} } % $
+    #1 $
+  }
+%    \end{macrocode}
+% \end{macro}
+% \begin{macro}{\@@_grab_dollardollar:w}
+%   And for the classical \TeX{} display structure.
+%    \begin{macrocode}
+\cs_new_protected:Npn \@@_grab_dollardollar:w % $$
+  #1 $$
+  {
+    \@@_process:nn { equation* } {#1} % $$
+    #1 $$
+  }
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}{\@@_grab_inline:w}
+%   Collect inline math content and deal with the need to move to math mode.
+%    \begin{macrocode}
+\cs_new_protected:Npn \@@_grab_inline:w % \(
+  #1 \)
+  {
+    \@@_process:nn { math } {#1}
+    $ #1 $
+    \bool_set_false:N \l_@@_collected_bool
+  }
+%    \end{macrocode}
```

Comment:
Two+ items -

One (obscure) bug affecting an edge case:
   Here one should use ` $ \relax #1 $ ` ,
   to guard against an empty ` \(\) ` (as in \@ensuredmath).

Two, a possible enhancement (covers an unlikely but not 
impossible case): 
   Add look-ahead for ` \m@th ` at start of ` \( `.

Two+  Note, not currently within this file:
      this look-ahead for ` \m@th ` is definitely needed
      within ` \ensuremath `, since a common usage is: 
     ` \ensuremath { \m@th ...`.

Reply:
Amendment: `\ensuremath` is in this file.

Reply:
I've covered both of these I hope.